Public/Src/Cache/ContentStore/Distributed/NuCache/ReplicatedRedisHashKey.cs

```diff
@@ -0,0 +1,210 @@
+    /// conditions that can happen in the event of failures. The assumption is that the data updated by this component is rarely updated
+    /// in a way that all operations must be successful for the data to remain valid. This is true for the current set of keys managed by this
+    /// component:
+    /// ClusterStateKey - this is an important key. Most update operations simply update the machines last heartbeat time. These can be lost for small periods of time with
+    /// no ill effects. 
+    ///     WARNING: Failures during new machine addition are not always handled properly by this strategy may result in recurring failures. New machine addition is very rare so
+    ///     the assumption that the likelihood of a failure during new machine addition will be essentially non-existent.
+    /// 
+    /// MasterLeaseKey - operations here are infrequent and transient failures would only mean cluster does not properly assign a master for a short period of time.
+    /// 
+    /// CheckpointsKey - most operations are reads. Transient update failures will simply mean the checkpoint is not updated so machines may end up going back to
+    /// slightly older point in time
+    /// </summary>
+    internal sealed class ReplicatedRedisHashKey
+    {
+        /// <summary>
+        /// Name of field in redis hash which indicates the version number of key. 
+        /// </summary>
+        private const string ReplicatedHashVersionNumber = nameof(ReplicatedHashVersionNumber);
+
+        private readonly string _key;
+        private readonly IReplicatedKeyHost _host;
+        private readonly IClock _clock;
+        private readonly RaidedRedisDatabase _redis;
+
+        private DateTime _lastMirrorTime;
+
+        /// <nodoc />
+        public ReplicatedRedisHashKey(string key, IReplicatedKeyHost host, IClock clock, RaidedRedisDatabase redis)
+        {
+            _clock = clock;
+            _redis = redis;
+            _key = key;
+            _host = host;
+        }
+
+        /// <summary>
+        /// Perform an operation against the redis hash using logic to ensure that result of operation comes from instance with latest updates and is resilient to data loss in one of the instances
+        /// </summary>
+        public async Task<Result<T>> UseReplicatedHashAsync<T>(OperationContext context, RedisOperation operation, Func<RedisBatch, string, Task<T>> addOperations, [CallerMemberName] string caller = null)
+        {
+            // Query to see which db has highest version number
+            (var primaryVersionedResult, var secondaryVersionedResult) = await _redis.ExecuteRaidedAsync(context, async redisDb =>
+            {
+                var operationResult = await redisDb.ExecuteBatchAsync(context,
+                    async batch =>
+                    {
+                        var versionTask = batch.AddOperation(_key, b => b.HashIncrementAsync(_key, nameof(ReplicatedHashVersionNumber)));

```

Comment:
Minor, but this may fail with unobserved task exception: if both addOpations and versionTask fail, then versionTask is unobserved.

Reply:
Will fix in subsequent changes.

---
In reply to: [337751017](https://github.com/microsoft/BuildXL/pull/1081#discussion_r337751017) [](ancestors = 337751017)